File: frontend/src/components/benchmark/content/dijkstra/implementations/dijkstra.php
<?php

class Dijkstra {
    private array $adjacency = [];
    private array $nodes = [];

    public function addEdge(int $u, int $v, int $weight): void {
        if (!isset($this->adjacency[$u])) {
            $this->adjacency[$u] = [];
        }
        if (!isset($this->adjacency[$v])) {
            $this->adjacency[$v] = [];
        }
        $this->adjacency[$u][] = [$v, $weight];
        $this->adjacency[$v][] = [$u, $weight];
        $this->nodes[$u] = true;
        $this->nodes[$v] = true;
    }

    public function addDirectedEdge(int $u, int $v, int $weight): void {
        if (!isset($this->adjacency[$u])) {
            $this->adjacency[$u] = [];
        }
        if (!isset($this->adjacency[$v])) {
            $this->adjacency[$v] = [];
        }
        $this->adjacency[$u][] = [$v, $weight];
        $this->nodes[$u] = true;
        $this->nodes[$v] = true;
    }

    public function buildFromEdges(array $edges, bool $directed = false): void {
        foreach ($edges as $edge) {
            if ($directed) {
                $this->addDirectedEdge($edge[0], $edge[1], $edge[2]);
            } else {
                $this->addEdge($edge[0], $edge[1], $edge[2]);
            }
        }
    }

    public function shortestPath(int $source): array {
        $dist = [];
        foreach (array_keys($this->nodes) as $node) {
            $dist[$node] = INF;
        }
        $dist[$source] = 0;

        $pq = [];
        $pq[] = [$source, 0];

        while (!empty($pq)) {
            usort($pq, function($a, $b) {
                return $a[1] - $b[1];
            });
            $current = array_shift($pq);
            $u = $current[0];
            $d = $current[1];

            if ($d != $dist[$u]) continue;

            if (!isset($this->adjacency[$u])) continue;

            foreach ($this->adjacency[$u] as $edge) {
                $v = $edge[0];
                $w = $edge[1];
                if ($dist[$u] + $w < $dist[$v]) {
                    $dist[$v] = $dist[$u] + $w;
                    $pq[] = [$v, $dist[$v]];
                }
            }
        }

        return $dist;
    }

    public function shortestPathWithPath(int $source, int $target): array {
        $dist = [];
        $parent = [];

        foreach (array_keys($this->nodes) as $node) {
            $dist[$node] = INF;
        }
        $dist[$source] = 0;
        $parent[$source] = -1;

        $pq = [];
        $pq[] = [$source, 0];

        while (!empty($pq)) {
            usort($pq, function($a, $b) {
                return $a[1] - $b[1];
            });
            $current = array_shift($pq);
            $u = $current[0];
            $d = $current[1];

            if ($d != $dist[$u]) continue;

            if ($u == $target) {
                $path = [];
                while ($u != -1) {
                    $path[] = $u;
                    $u = $parent[$u];
                }
                return array_reverse($path);
            }

            if (!isset($this->adjacency[$u])) continue;

            foreach ($this->adjacency[$u] as $edge) {
                $v = $edge[0];
                $w = $edge[1];
                if ($dist[$u] + $w < $dist[$v]) {
                    $dist[$v] = $dist[$u] + $w;
                    $parent[$v] = $u;
                    $pq[] = [$v, $dist[$v]];
                }
            }
        }

        return [];
    }

    public function allShortestPaths(int $source): array {
        $paths = [];
        $dist = [];

        foreach (array_keys($this->nodes) as $node) {
            $dist[$node] = INF;
        }
        $dist[$source] = 0;
        $paths[$source] = [$source];

        $pq = [];
        $pq[] = [$source, 0];

        while (!empty($pq)) {
            usort($pq, function($a, $b) {
                return $a[1] - $b[1];
            });
            $current = array_shift($pq);
            $u = $current[0];
            $d = $current[1];

            if ($d != $dist[$u]) continue;

            if (!isset($this->adjacency[$u])) continue;

            foreach ($this->adjacency[$u] as $edge) {
                $v = $edge[0];
                $w = $edge[1];
                if ($dist[$u] + $w < $dist[$v]) {
                    $dist[$v] = $dist[$u] + $w;
                    $paths[$v] = array_merge($paths[$u], [$v]);
                    $pq[] = [$v, $dist[$v]];
                }
            }
        }

        return $paths;
    }
}

?>

File: frontend/src/components/benchmark/content/dijkstra/implementations/dijkstra.java
import java.util.*;

public class Dijkstra {
    private Map<Integer, List<int[]>> adjacency = new HashMap<>();
    private Set<Integer> nodes = new HashSet<>();

    public void addEdge(int u, int v, int weight) {
        adjacency.putIfAbsent(u, new ArrayList<>());
        adjacency.putIfAbsent(v, new ArrayList<>());
        adjacency.get(u).add(new int[]{v, weight});
        adjacency.get(v).add(new int[]{u, weight});
        nodes.add(u);
        nodes.add(v);
    }

    public void addDirectedEdge(int u, int v, int weight) {
        adjacency.putIfAbsent(u, new ArrayList<>());
        adjacency.putIfAbsent(v, new ArrayList<>());
        adjacency.get(u).add(new int[]{v, weight});
        nodes.add(u);
        nodes.add(v);
    }

    public void buildFromEdges(List<int[]> edges, boolean directed) {
        for (int[] edge : edges) {
            if (directed) {
                addDirectedEdge(edge[0], edge[1], edge[2]);
            } else {
                addEdge(edge[0], edge[1], edge[2]);
            }
        }
    }

    public Map<Integer, Integer> shortestPath(int source) {
        Map<Integer, Integer> dist = new HashMap<>();
        for (int node : nodes) {
            dist.put(node, Integer.MAX_VALUE);
        }
        dist.put(source, 0);

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
        pq.offer(new int[]{source, 0});

        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int u = current[0];
            int d = current[1];

            if (d != dist.get(u)) continue;

            if (!adjacency.containsKey(u)) continue;

            for (int[] edge : adjacency.get(u)) {
                int v = edge[0];
                int w = edge[1];
                if (dist.get(u) + w < dist.get(v)) {
                    dist.put(v, dist.get(u) + w);
                    pq.offer(new int[]{v, dist.get(v)});
                }
            }
        }

        return dist;
    }

    public List<Integer> shortestPathWithPath(int source, int target) {
        Map<Integer, Integer> dist = new HashMap<>();
        Map<Integer, Integer> parent = new HashMap<>();

        for (int node : nodes) {
            dist.put(node, Integer.MAX_VALUE);
        }
        dist.put(source, 0);
        parent.put(source, -1);

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
        pq.offer(new int[]{source, 0});

        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int u = current[0];
            int d = current[1];

            if (d != dist.get(u)) continue;

            if (u == target) {
                List<Integer> path = new ArrayList<>();
                while (u != -1) {
                    path.add(u);
                    u = parent.get(u);
                }
                Collections.reverse(path);
                return path;
            }

            if (!adjacency.containsKey(u)) continue;

            for (int[] edge : adjacency.get(u)) {
                int v = edge[0];
                int w = edge[1];
                if (dist.get(u) + w < dist.get(v)) {
                    dist.put(v, dist.get(u) + w);
                    parent.put(v, u);
                    pq.offer(new int[]{v, dist.get(v)});
                }
            }
        }

        return new ArrayList<>();
    }

    public Map<Integer, List<Integer>> allShortestPaths(int source) {
        Map<Integer, List<Integer>> paths = new HashMap<>();
        Map<Integer, Integer> dist = new HashMap<>();

        for (int node : nodes) {
            dist.put(node, Integer.MAX_VALUE);
        }
        dist.put(source, 0);
        paths.put(source, new ArrayList<>(Arrays.asList(source)));

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
        pq.offer(new int[]{source, 0});

        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int u = current[0];
            int d = current[1];

            if (d != dist.get(u)) continue;

            if (!adjacency.containsKey(u)) continue;

            for (int[] edge : adjacency.get(u)) {
                int v = edge[0];
                int w = edge[1];
                if (dist.get(u) + w < dist.get(v)) {
                    dist.put(v, dist.get(u) + w);
                    paths.put(v, new ArrayList<>(paths.get(u)));
                    paths.get(v).add(v);
                    pq.offer(new int[]{v, dist.get(v)});
                }
            }
        }

        return paths;
    }
}